# TEAM 4 CLASS PROJECT PYTHON NOTEBOOK

Columbia University APAN 5310 - SQL & Relational Databases (Fall 2018)

Prepared by Erman Gao, Li Lai, Yuqing Lei, Yike He. Posted on Dec 3, 2018

#### First, import necessary packages

In [2]:
import pandas as pd
from sqlalchemy import create_engine

## Create connection between Python and PostgreSQL

In [3]:
# Pass the connection string to a variable, conn_url
conn_url = 'postgresql://postgres:pwd4APAN5310@35.231.194.11/team4'

In [4]:
# Create an engine that connects to PostgreSQL server
engine = create_engine(conn_url)

In [5]:
# Establish a connection
connection = engine.connect()

In [6]:
# Load the csv file
df = pd.read_csv( "DOHMH_New_York_City_Restaurant_Inspection_Results.csv", sep= ',', encoding='latin-1')

In [7]:
# Create a random sample datafram
df = df.sample(frac=0.25, replace=False)

In [95]:
# Read the table 
df.head()

CAMIS             DBA       BORO BUILDING         STREET  ZIPCODE  \
229939  50043978        KINGSLEY  MANHATTAN      190       AVENUE B  10009.0   
30699   41528976  FAMILY KITCHEN   BROOKLYN     1709  FOSTER AVENUE  11230.0   
229942  50083487        MAMAGYRO  MANHATTAN      165     E 106TH ST  10029.0   
207701  50082836      CHINA KING     QUEENS    10119    QUEENS BLVD  11375.0   
65857   50010310          HANABI  MANHATTAN     1450        2ND AVE  10021.0   

             PHONE CUISINE DESCRIPTION INSPECTION DATE  \
229939  2126744500            American        3/3/2018   
30699   7188590283             Chinese       2/15/2018   
229942  2124101111               Greek       11/7/2018   
207701  7189978000       Chinese/Cuban       11/5/2018   
65857   2125701888            Japanese       2/16/2017   

                                                 ACTION VIOLATION CODE  \
229939  Violations were cited in the following area(s).            09C   
30699   Violations were cited in the following area(s).            06C   
229942                 Establishment re-opened by DOHMH            06C   
207701  Violations were cited in the following area(s).            10F   
65857   Violations were cited in the following area(s).            04C   

                                    VIOLATION DESCRIPTION CRITICAL FLAG  \
229939      Food contact surface not properly maintained.  Not Critical   
30699   Food not protected from potential source of co...      Critical   
229942  Food not protected from potential source of co...      Critical   
207701  Non-food contact surface improperly constructe...  Not Critical   
65857   Food worker does not use proper utensil to eli...      Critical   

        SCORE GRADE                                  INSPECTION TYPE  
229939   12.0     A                 Cycle Inspection / Re-inspection  
30699     6.0     A            Cycle Inspection / Initial Inspection  
229942    7.0     Z  Pre-permit (Operational) / Reopening Inspection  
207701   10.0     A    Pre-permit (Operational) / Initial Inspection  
65857    27.0   NaN            Cycle Inspection / Initial Inspection

In [96]:
# Look at the basic information
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 60139 entries, 229939 to 202290
Data columns (total 16 columns):
CAMIS                    60139 non-null int64
DBA                      60139 non-null object
BORO                     60139 non-null object
BUILDING                 60107 non-null object
STREET                   60139 non-null object
ZIPCODE                  59245 non-null float64
PHONE                    60139 non-null object
CUISINE DESCRIPTION      60139 non-null object
INSPECTION DATE          60139 non-null object
ACTION                   60139 non-null object
VIOLATION CODE           59410 non-null object
VIOLATION DESCRIPTION    58961 non-null object
CRITICAL FLAG            60139 non-null object
SCORE                    57540 non-null float64
GRADE                    30512 non-null object
INSPECTION TYPE          60139 non-null object
dtypes: float64(2), int64(1), object(13)
memory usage: 7.8+ MB


#### After creating the random sample dataframe, there is 60,139 rows and 16 columns in total.

# Extract, Transform and Load (ETL)

For every single dataframe, we dropped duplicates before creating an “ID column” to ensure that every primary key represents unique information. And we also renamed the columns in every dataframe so that when pushing the information to SQL database, every column can find its corresponding column. 

We prepared 10 dataframes for each of the normalized table by selecting the relevant columns.

In [10]:
# Populate the violations table
violations_df = df[['VIOLATION CODE','VIOLATION DESCRIPTION']]
violations_df = violations_df.drop_duplicates()
violations_df.insert( 0 , 'violation_id' , range( 1 , 1 + len(violations_df)))
violations_df = violations_df.rename(columns = {'VIOLATION CODE':'violation_code','VIOLATION DESCRIPTION':'violation_description'})
violations_df.to_sql(name= 'violations' , con=engine, if_exists= 'append' , index= False )

In [11]:
# Populate the inspections table
inspection_df = pd.DataFrame(df['INSPECTION TYPE'].unique(), columns=[ 'inspection_type' ])
inspection_df.insert( 0 , 'inspection_id' , range( 1 , 1 + len(inspection_df)))
inspection_df.to_sql(name='inspections', con=engine, if_exists='append', index=False)

In [12]:
# Populate the actions table
action_df = pd.DataFrame(df['ACTION'].unique(), columns=[ 'action' ])
action_df.insert( 0 , 'action_id' , range( 1 , 1 + len(action_df)))
action_df.to_sql(name='actions', con=engine, if_exists='append', index=False)

In [13]:
# Populate the critical_flags table
critical_flag_df = pd.DataFrame(df['CRITICAL FLAG'].unique(), columns=['critical_flag'])
critical_flag_df.insert( 0 , 'flag_id' , range( 1 , 1 + len(critical_flag_df)))
critical_flag_df.to_sql(name= 'critical_flags' , con=engine, if_exists= 'append' , index= False )

In [14]:
# Populate the grades table
grade_df = pd.DataFrame(df['GRADE'].unique(), columns=['grade'])
grade_df.insert( 0 , 'grade_id' , range( 1 , 1 + len(grade_df)))
grade_df.to_sql(name= 'grades' , con=engine, if_exists= 'append' , index= False )

In [15]:
# Populate the boroughs table
borough = df['BORO'].unique()
borough_df = pd.DataFrame(borough, columns=['borough'])
borough_df.insert( 0 , 'borough_id' , range( 1 , 1 + len(borough_df)))
borough_df.to_sql(name= 'boroughs' , con=engine, if_exists= 'append' , index= False )

In [16]:
# Populate the zipcodes table
zipcode = df['ZIPCODE'].unique()
zipcode_df = pd.DataFrame(zipcode, columns=['zipcode'])
zipcode_df.insert( 0 , 'zipcode_id' , range( 1 , 1 + len(zipcode_df)))
zipcode_df.to_sql(name= 'zipcodes' , con=engine, if_exists= 'append' , index= False )

In [18]:
# Populate the cuisine_type table
cuisine_type_df = pd.DataFrame(df['CUISINE DESCRIPTION'].unique(), columns=[ 'cuisine_type' ])
cuisine_type_df.insert( 0 , 'cuisine_type_id' , range( 1 , 1 + len(cuisine_type_df)))
cuisine_type_df.to_sql(name= 'cuisine_types' , con=engine, if_exists= 'append' , index= False )

In [19]:
# Populate the restaurants table
restaurants = df[['CAMIS','DBA','BORO','CUISINE DESCRIPTION','BUILDING','STREET','PHONE','ZIPCODE']]
restaurants = restaurants.drop_duplicates()
restaurants.columns = ['restaurant_id','name','borough','cuisine_type','building','street','phone','zipcode']
df2 = pd.merge (restaurants, zipcode_df, on ='zipcode')
df3 = pd.merge (df2, borough_df, on = 'borough')
df4 = pd.merge (df3, cuisine_type_df, on = 'cuisine_type')
restaurants_df = df4[['restaurant_id','name','building','street','phone','zipcode_id','borough_id','cuisine_type_id']]
restaurants_df.to_sql(name='restaurants', con=engine, if_exists='append', index=False)

In [20]:
# Populate the inspection_records table
inspection_records = df[['CAMIS','INSPECTION TYPE','ACTION','VIOLATION CODE','CRITICAL FLAG','INSPECTION DATE','GRADE','SCORE']]
inspection_records = inspection_records.drop_duplicates()
inspection_records.columns = ['restaurant_id','inspection_type','action','violation_code','critical_flag','inspection_date','grade','score']
df5 = pd.merge (inspection_records, inspection_df, on = 'inspection_type')
df6 = pd.merge (df5, action_df, on = 'action')
df7 = pd.merge (df6, violations_df, on = 'violation_code')
df8 = pd.merge (df7, critical_flag_df, on = 'critical_flag')
df9 = pd.merge (df8, grade_df, on = 'grade')
inspection_records_df = df9[['restaurant_id','inspection_id','action_id','violation_id','flag_id','inspection_date','grade_id','score']]
inspection_records_df.insert( 0 , 'inspection_record_id' , range( 1 , 1 + len(inspection_records_df)))
inspection_records_df.to_sql(name='inspection_records', con=engine, if_exists='append', index=False)

When dealing with creating a dataframe for the large tables that contain different unique ID information, we merged the dataframes that we created earlier that contain their own unique ID column with the current dataframe we are creating to match the unique IDs all together. 